docs/binary-transparency-for-artifact-registries.md

```diff
@@ -0,0 +1,160 @@
+# Binary Transparency for Artifact Registries
+
+Authors: [Hayden Blauzvern (Google)](https://github.com/haydentherapper)
+
+Last updated: September 2024
+
+This document defines Binary Transparency for artifacts registries to mitigate the risk of registry compromise by external attackers or insiders.
+
+## What is Binary Transparency?
+
+[Binary Transparency](https://binary.transparency.dev/) (BT) is a security solution that adds cryptographic immutability and discoverability to package identifiers and metadata, which in turn prevents tampering of existing versions and supports detection of unauthorized new versions.
+
+BT is backed by a cryptographic data structure that is immutable and append-only. When publicly accessible, this structure becomes a transparency log, providing transparency into every committed decision present in the log. Each entry into the log receives an inclusion proof, a cryptographically verifiable confirmation of the inclusion of the entry, which is checked during artifact verification.
+
+In the context of artifact registries, BT provides an assurance that all users always receive the same binary for a given package identifier, because users require the registry to transparently commit to all binaries it serves. If a registry is compromised or compelled to serve a different binary, then it must do so for all of its users and for all time, and because this decision must be recorded in a public transparency log, the malicious action can be audited and detected.
+
+## Applying Binary Transparency to Artifact Registries
+
+BT applies to any package distribution method using logically immutable package identifiers[^1]. This includes most OSS package ecosystems, such as Go, PyPI (Python), Cargo (Rust), npm (JavaScript), Maven Central (Java), and CI/CD registries such as GitHub Releases[^2]. In other words, users identify and fetch a package by `<package_name>`, `<version_id>`, and `<filename>` (or similar), referred to as a package identifier, and receive back some artifact. Logically, this process should *always* and *globally* return the same artifact[^3]. However, most ecosystems lack sufficient cryptographic security controls to ensure that this always occurs, making them susceptible to tampering by compromised registries, mirrors, or distribution channels. This is where BT fits in.
+
+BT makes package identifiers immutable and transparent:
+
+* A package registry will publish a commitment that a specific package identifier maps to a cryptographic digest. This commitment creates an immutable mapping between the identifier and digest. The mapping can only be done per package identifier and can never be updated or deleted.  
```

Comment:
Most of the ecosystems listed above have some form of package deletion. Many provide a way for a user to request one of the packages be deleted, in a self-service manner, under various safeguards. Even the ones that don't will remove an artifact when compelled by a court.  Generally, if they allow deletion they also allow there to be a new owner who is allowed to publish new packages at the same version number. Unfortunately, these corner cases cannot be ignored. How does the system accommodate this reality?

Reply:
Great question! In one of the [footnotes](https://github.com/ossf/wg-securing-software-repos/pull/48/files#diff-8e0542c5388ca2387892b0ed56a55b81130d24fd3a094c59988be53a22021accR158), I touch on if the ecosystem supports yanking a package. Supporting package deletion wouldn't be at odds with binary transparency, _as long as_ the ecosystem wouldn't allow package identifier reuse, e.g. if `package@v1.2.3` is deleted from the registry, then the next package version **must** be `v1.2.4` or higher (or more accurately, a version that's never been used before, BT is not opinionated on ever-increasing versions, that's an ecosystem decision).

Lockfile support in an ecosystem also should require this to be true, otherwise yanking and recreating a package at a given version would break existing lockfiles.

A question for the registries, do you allow version reuse? At a glance:

* [npm does not allow version reuse](https://docs.npmjs.com/policies/unpublish#considerations)
* [PyPI does not allow version reuse if a file was deleted](https://pypi.org/help/#file-name-reuse)
* [Maven Central doesn't support deletion](https://central.sonatype.org/faq/can-i-change-a-component/)
* [RubyGems supports yanking, but unspecified on version reuse](https://guides.rubygems.org/removing-a-published-gem/)
* [NuGet supports unlisting, but not deleting iiuc](https://learn.microsoft.com/en-us/nuget/nuget-org/policies/deleting-packages)
* Go catches version reuse through its [proxy](https://sum.golang.org/) since it runs a binary transparency log. For example, if a tag is deleted and recreated pointing to a different commit hash, the Go proxy will throw an error.

Reply:
To tack onto what @haydentherapper said: PyPI's behavior is subtle in that *filenames* are always unique and immutable on PyPI, but *releases* themselves are not. In other words: a project `foo` that gets deleted or turned over to a new user can't overwrite `foo-1.2.3.tar.gz` if that distribution file was already uploaded by a previous maintainer, but the new maintainer *might* be able to upload `foo-1.2.3-py3-none-any.whl` or similar if no wheel was previously uploaded for that version of `foo`. 

In practice this means that resolving `foo==1.2.3` isn't guaranteed to be stable for a given host, since the new maintainer can always upload a new (unique) file to the old version that's more specific/matches the host's target configuration more precisely.